src/server/http/xsrf.js

```diff
@@ -4,18 +4,37 @@ export default function (kbnServer, server, config) {
   const disabled = config.get('server.xsrf.disableProtection');
   const versionHeader = 'kbn-version';
   const xsrfHeader = 'kbn-xsrf';
+  const contentTypeHeader = 'content-type';
+  const allowedRequestMediaTypes = ['application/json', 'application/x-ndjson'];
 
   server.ext('onPostAuth', function (req, reply) {
     if (disabled) {
       return reply.continue();
     }
 
     const isSafeMethod = req.method === 'get' || req.method === 'head';
-    const hasVersionHeader = versionHeader in req.headers;
-    const hasXsrfHeader = xsrfHeader in req.headers;
+    if (!isSafeMethod) {
+      const hasVersionHeader = versionHeader in req.headers;
+      const hasXsrfHeader = xsrfHeader in req.headers;
 
-    if (!isSafeMethod && !hasVersionHeader && !hasXsrfHeader) {
-      return reply(badRequest(`Request must contain an ${xsrfHeader} header`));
+      if (hasXsrfHeader || hasVersionHeader) {
+        server.log(
+          ['warning', 'deprecation'],
+          `Both ${versionHeader} and ${xsrfHeader} headers are deprecated and not used for CSRF protection anymore. ` +
+          `Please use proper ${contentTypeHeader} header instead.`
+        );
+      }
+
+      const hasContentTypeHeader = contentTypeHeader in req.headers;
+      // Since we're only interested in media type let's extract it from the content type ("media type [;parameter]")
+      // and leave leave off parameter portion (e.g. charset) of the header.
+      const hasAllowedMediaType = hasContentTypeHeader && allowedRequestMediaTypes.includes(
+        req.headers[contentTypeHeader].split(';')[0]
+      );
+
+      if (!hasAllowedMediaType && !hasVersionHeader && !hasXsrfHeader) {
+        return reply(badRequest(`Request must contain a proper ${contentTypeHeader} header.`));
```

Comment:
Can you expand on this error message to include the two possible valid content-types?  If we can guide people to the solution without them having to guess or jump to docs, then that's a good thing.